singularity-eos/eos/singularity_eos.cpp

```diff
@@ -343,327 +351,512 @@ int get_sg_eos( // sizing information
-        sie_v(i) = s_is_inp ? sie_v(i) : 0.0;
-        temp_v(i) = t_is_inp ? temp_v(i) : 0.0;
-        bmod_v(i) = 0.0;
-        cv_v(i) = 0.0;
-        dpde_v(i) = 0.0;
-        Real vol_sum = 0.0;
-        for (int m = 0; m < npte; ++m) {
-          int mid = pte_mats(tid, m);
-          int midx = pte_idxs(tid, m);
-          Real rho = mass_pte(tid, m) / vfrac_pte(tid, m);
-          Real spie = sie_pte(tid, m);
-          Real press_mat = press_v(i);
-          Real T = temp_v(i) * ev2k;
-          Real cv_mat = 0.0;
-          Real bmod_mat = 0.0;
-          eos_v(mid).FillEos(rho, T, spie, press_mat, cv_mat, bmod_mat, output, cache);
-          frac_sie_v(i, midx) = spie * frac_mass_v(i, midx);
-          frac_vol_v(i, midx) = frac_mass_v(i, midx) / rho;
-          vol_sum += frac_vol_v(i, midx);
-          temp_v(i) += t_is_inp ? 0.0 : frac_vol_v(i, midx) / ev2k * T;
-          press_v(i) += p_is_inp ? 0.0 : frac_vol_v(i, midx) * press_mat;
-          // optional cv and bmod
-          const Real dpde_mat =
-              rho * eos_v(mid).GruneisenParamFromDensityInternalEnergy(rho, spie, cache);
-          sie_v(i) += s_is_inp ? 0.0 : frac_sie_v(i, midx);
-          bmod_v(i) += frac_vol_v(i, midx) * bmod_mat;
-          cv_v(i) += frac_mass_v(i, midx) * cv_mat;
-          dpde_v(i) += frac_vol_v(i, midx) * dpde_mat;
-          if (do_frac_bmod) {
-            frac_bmod_v(i, midx) = bmod_mat;
-          }
-          if (do_frac_cv) {
-            frac_cv_v(i, midx) = cv_mat;
-          }
-          if (do_frac_dpde) {
-            frac_dpde_v(i, midx) = dpde_mat;
-          }
-        }
-        press_v(i) /= p_is_inp ? 1.0 : vol_sum;
-        sie_v(i) /= s_is_inp ? 1.0 : mass_sum;
-        temp_v(i) /= t_is_inp ? 1.0 : vol_sum;
-        bmod_v(i) /= vol_sum;
-        cv_v(i) *= ev2k / mass_sum;
-        dpde_v(i) /= vol_sum;
-        spvol_v(i) = r_is_inp ? spvol_v(i) : vol_sum / mass_sum;
-        pmax_v(i) = press_v(i) > pmax_v(i) ? press_v(i) : pmax_v(i);
-        // if failed, increment return value by 1
-        // use if rather than ternary to reduce contention
-        if (!res_) {
-          res() += 1;
-        }
-        // release token
-        tokens.release(token);
-      });
+  switch(input_int) {
+  case -3:
+    // T-rho input
+    // set frac_vol = 1/nmat
+    // set rho_i = nmat / spvol * frac_mass_i
+    // iterate PTE solver to obtain internal energies
+    // that results in the input T
+    pte_solver_scratch_size = PTESolverFixedTRequiredScratch(nmat);
+    solver_scratch = ScratchV<double>("PTE::scratch solver", scratch_size,
+				      pte_solver_scratch_size);
+    portableFor("PTE::solve (rho,T) input", 0, ncell,
+		PORTABLE_LAMBDA(const int &iloop)
+    {
+      // cell offset
+      const int i{offsets_v(iloop) - 1};
+      // get "thread-id" like thing with optimization
+      // for small loops
+      const int32_t token{tokens.acquire()};
+      const int32_t tid{small_loop ? iloop : token};
```

Comment:
Can the tokenization stuff for Kokkos be abstracted outside the switch statement?

Reply:
They can't, `iloop` is the lambda variable.

Reply:
ok :+1: